src/hotspot/share/services/finalizerService.cpp

```diff
@@ -0,0 +1,488 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#if INCLUDE_MANAGEMENT
+#include "classfile/classLoaderDataGraph.inline.hpp"
+#include "logging/log.hpp"
+#include "runtime/atomic.hpp"
+#include "runtime/handles.hpp"
+#include "runtime/interfaceSupport.inline.hpp"
+#include "runtime/javaCalls.hpp"
+#include "runtime/mutexLocker.hpp"
+#include "runtime/synchronizer.hpp"
+#include "runtime/thread.inline.hpp"
+#include "runtime/vm_version.hpp"
+#include "services/finalizerService.hpp"
+#include "utilities/concurrentHashTableTasks.inline.hpp"
+#include "utilities/debug.hpp"
+
+FinalizerEntry::FinalizerEntry(const InstanceKlass* ik) :
+    _ik(ik),
+    _objects_on_heap(0),
+    _total_finalizers_run(0) {}
+
+const InstanceKlass* FinalizerEntry::klass() const {
+  return _ik;
+}
+
+uint64_t FinalizerEntry::objects_on_heap() const {
+  return Atomic::load(&_objects_on_heap);
+}
+
+uint64_t FinalizerEntry::total_finalizers_run() const {
+  return Atomic::load(&_total_finalizers_run);
+}
+
+template <uint64_t op(uint64_t)>
+static inline void set_atomic(volatile uint64_t* volatile dest) {
+  assert(VM_Version::supports_cx8(), "invariant");
+  uint64_t compare;
+  uint64_t exchange;
+  do {
+    compare = *dest;
+    exchange = op(compare);
+  } while (Atomic::cmpxchg(dest, compare, exchange) != compare);
+}
+
+static inline uint64_t inc(uint64_t value) {
+  return value + 1;
+}
+
+void FinalizerEntry::on_register() {
+  set_atomic<inc>(&_objects_on_heap);
+}
+
+static inline uint64_t dec(uint64_t value) {
+  assert(value > 0, "invariant");
+  return value - 1;
+}
+
+void FinalizerEntry::on_complete() {
+  set_atomic<inc>(&_total_finalizers_run);
+  set_atomic<dec>(&_objects_on_heap);
+}
+
+static constexpr const size_t DEFAULT_TABLE_SIZE = 2048;
+// 2^24 is max size, like StringTable.
+static constexpr const size_t MAX_SIZE = 24;
+// If a chain gets to 50, something might be wrong
+static constexpr const size_t REHASH_LEN = 50;
+static constexpr const double PREF_AVG_LIST_LEN = 8.0;
+
+static size_t _table_size = 0;
+static volatile uint64_t _entries = 0;
+static volatile uint64_t _count = 0;
+static volatile bool _has_work = 0;
+static volatile bool _needs_rehashing = false;
+static volatile bool _has_items_to_clean = false;
+
+static inline void reset_has_items_to_clean() {
+  Atomic::store(&_has_items_to_clean, false);
+}
+
+static inline void set_has_items_to_clean() {
+  Atomic::store(&_has_items_to_clean, true);
+}
+
+static inline bool has_items_to_clean() {
+  return Atomic::load(&_has_items_to_clean);
+}
+
+static inline void added() {
+  set_atomic<inc>(&_count);
+}
+
+static inline void removed() {
+  set_atomic<dec>(&_count);
+}
+
+static inline uintx hash_function(const InstanceKlass* ik) {
+  assert(ik != nullptr, "invariant");
+  return primitive_hash(ik);
```

Comment:
Thank you for pointing that out. I had assumed this to be a part of the syntactic contract for using the ConcurrentHashTable. My wrong assumption was because the SymbolTable (which I used as a model) seemed to pass a "rehash_warning" bool to the accessor APIs (get, insert). However, looking more closely at the signatures in ConcurrentHashTable, that bool is a "grow_hint", not "rehash" specifically. Thanks again; I will remove the rehash support code.